platforms/allwinner-d1/core/src/ccu.rs

```diff
@@ -0,0 +1,260 @@
+            w.div_cfg().variant(1);
+            w
+        });
+        sdelay(1);
+
+        /* Disable pll gating */
+        self.ccu
+            .pll_cpu_ctrl
+            .modify(|_, w| w.pll_output_gate().disable());
+
+        /* Enable pll ldo */
+        self.ccu.pll_cpu_ctrl.modify(|_, w| w.pll_ldo_en().enable());
+        sdelay(5);
+
+        /* Set default clk to 1008mhz */
+        self.ccu.pll_cpu_ctrl.modify(|r, w| {
+            // undocumented part of register is cleared by xboot
+            unsafe { w.bits(r.bits() & !(0x3 << 16)) };
+            w.pll_m().variant(0);
+            w.pll_n().variant(41);
+            w
+        });
+
+        /* Lock enable */
+        self.ccu
+            .pll_cpu_ctrl
+            .modify(|_, w| w.lock_enable().enable());
+
+        /* Enable pll */
+        self.ccu.pll_cpu_ctrl.modify(|_, w| w.pll_en().enable());
+
+        /* Wait pll stable */
+        while self.ccu.pll_cpu_ctrl.read().lock().bit_is_clear() {
+            core::hint::spin_loop();
+        }
+        sdelay(20);
+
+        /* Enable pll gating */
+        self.ccu
+            .pll_cpu_ctrl
+            .modify(|_, w| w.pll_output_gate().enable());
+
+        /* Lock disable */
+        self.ccu
+            .pll_cpu_ctrl
+            .modify(|_, w| w.lock_enable().disable());
+        sdelay(1);
+
+        /* Set and change cpu clk src */
+        self.ccu.riscv_clk.modify(|_, w| {
+            w.clk_src_sel().pll_cpu();
+            w.axi_div_cfg().variant(1);
+            w.div_cfg().variant(0);
+            w
+        });
+        sdelay(1);
+    }
+
+    fn set_pll_periph0(&mut self) {
+        /* Periph0 has been enabled */
+        if self.ccu.pll_peri_ctrl.read().pll_en().bit_is_set() {
+            return;
+        }
+
+        /* Change psi src to osc24m */
+        self.ccu.psi_clk.modify(|_, w| w.clk_src_sel().hosc());
+
+        /* Set default val */
+        self.ccu.pll_peri_ctrl.write(|w| w.pll_n().variant(0x63));
+
+        /* Lock enable */
+        self.ccu
+            .pll_peri_ctrl
+            .modify(|_, w| w.lock_enable().enable());
+
+        /* Enabe pll 600m(1x) 1200m(2x) */
+        self.ccu.pll_peri_ctrl.modify(|_, w| w.pll_en().enable());
+
+        /* Wait pll stable */
+        while self.ccu.pll_peri_ctrl.read().lock().bit_is_clear() {
+            core::hint::spin_loop();
+        }
+        sdelay(20);
+
+        /* Lock disable */
+        self.ccu
+            .pll_peri_ctrl
+            .modify(|_, w| w.lock_enable().disable());
+    }
+
+    fn set_ahb(&mut self) {
+        self.ccu
+            .psi_clk
+            .write(|w| w.factor_m().variant(2).factor_n().n1());
+        self.ccu
+            .psi_clk
+            .modify(|_, w| w.clk_src_sel().pll_peri_1x());
+        sdelay(1);
+    }
+
+    fn set_apb(&mut self) {
+        self.ccu.apb_clk[0].write(|w| w.factor_m().variant(2).factor_n().n2());
+        self.ccu.apb_clk[0].modify(|_, w| w.clk_src_sel().pll_peri_1x());
```

Comment:
similarly, it might be nice to have a comment saying that we do this in a pair of separate writes to the register because xboot does that.